contracts/permissionless/OrderbookReserve.sol

```diff
@@ -0,0 +1,870 @@
+
+        require(removeOrder(list, maker, isEthToToken ? ETH_TOKEN_ADDRESS : token, orderId));
+
+        //funds go back to makers account
+        makerFunds[maker][isEthToToken ? ETH_TOKEN_ADDRESS : token] += orderData.srcAmount;
+
+        OrderCanceled(maker, isEthToToken, orderId, orderData.srcAmount, orderData.dstAmount);
+
+        return true;
+    }
+
+    ///@param maker is the maker of this order
+    ///@param isEthToToken which order type the maker is updating / adding
+    ///@param srcAmount is the orders src amount (token or ETH) could be negative if funds are released.
+    function bindOrderFunds(address maker, bool isEthToToken, int256 srcAmount)
+        internal
+        returns(bool)
+    {
+        address fundsAddress = isEthToToken ? ETH_TOKEN_ADDRESS : token;
+
+        if (srcAmount < 0) {
+            makerFunds[maker][fundsAddress] += uint(-srcAmount);
+        } else {
+            require(makerFunds[maker][fundsAddress] >= uint(srcAmount));
+            makerFunds[maker][fundsAddress] -= uint(srcAmount);
+        }
+
+        return true;
+    }
+
+    function getOrderData(OrderListInterface list, uint32 orderId) internal view returns (OrderData data)
+    {
+        uint32 prevId;
+        (data.maker, data.srcAmount, data.dstAmount, prevId, data.nextId) = list.getOrderDetails(orderId);
+        data.isLastOrder = (data.nextId == orderListTailId);
+    }
+
+    function bindOrderStakes(address maker, int stakeAmount) internal returns(bool) {
+
+        MakerKncAmounts storage amounts = makerKnc[maker];
+
+        require(amounts.freeKnc >= stakeAmount);
+
+        if (stakeAmount >= 0) {
+            amounts.freeKnc -= uint128(stakeAmount);
+            amounts.kncOnStake += uint128(stakeAmount);
+        } else {
+            amounts.freeKnc += uint128(-stakeAmount);
+            require(amounts.kncOnStake - uint128(-stakeAmount) < amounts.kncOnStake);
+            amounts.kncOnStake -= uint128(-stakeAmount);
+        }
+
+        return true;
+    }
+
+    ///@dev if burnAmount is 0 we only release stakes.
+    ///@dev if burnAmount == stakedAmount. all staked amount will be burned. so no knc returned to maker
+    function handleOrderStakes(address maker, uint stakedAmount, uint burnAmount) internal returns(bool) {
+
+        MakerKncAmounts storage amounts = makerKnc[maker];
+
+        //if knc rate had a big change. Previous stakes might not be enough. but can still take order.
+        if (amounts.kncOnStake < uint128(stakedAmount)) stakedAmount = amounts.kncOnStake;
+        if (burnAmount > stakedAmount) burnAmount = stakedAmount;
+
+        amounts.kncOnStake -= uint128(stakedAmount);
+        amounts.freeKnc += uint128(stakedAmount - burnAmount);
+
+        return true;
+    }
+
+    ///@dev funds are valid only when required knc amount can be staked for this order.
+    function validateAddOrder(address maker, bool isEthToToken, uint128 srcAmount, uint128 dstAmount)
```

Comment:
the name validate seems miss-leading as I expect a function to just check some stuff. maybe "secure" instead (if this function stays)?

Reply:
secure is also misleading IMO. other options?
I validate enough funds and enough KNC stake.